added in Socket code

t-keazirian opened the pull request:
~~Created a Socket~~
~~- Postman is working with both a GET/POST request and is echoing back the message that was sent:~~
~~- The connection stays alive~~
~~- Behavior that is unexpected (I think) is that it doesn't seem to ever break out of the outer `while` loop - it doesn't ever get to lines 70 or 71~~

Updating because the code has changed for the single-message implementation:

Created a Socket:

- Server can echo back one message from the client
- Used postman to send/connect

Editing to add:
![Screen Shot 2022-05-26 at 4 58 53 PM](https://user-images.githubusercontent.com/73620216/170586949-29e439e2-71d2-44e4-b895-1b9e0250345e.png)

Ian and I worked together to figure out I was not sending the correct response to Postman.  I fixed it and now have no errors (see pic above).

t-keazirian commented on TKeazirian.HTTPServer/Program.cs:
```diff
@@ -1,11 +1,79 @@
+            listener.Bind(localEndPoint);
+            listener.Listen(10);
+
+            while (true)
```
So if I remove the outer `while` loop, I am able to send one message and the server will echo it.  Which is the first [card](https://trello.com/c/IqnPvVzq/17-server-echoes-one-message).  However, if we want the connection to persist to send multiple messages, that is what that `while` loop is there for (I *think*)...

tpetersen0308 replied:
Sounds to me like you might be close to ready to open a PR for the "server can echo a message" story

t-keazirian commented on TKeazirian.HTTPServer/Program.cs:
```diff
@@ -1,11 +1,79 @@
+                    {
+                        break;
+                    }
+                }
```
Yes that is my understanding.

tpetersen0308 replied:
Wondering if there's a way to read the data out without looping, so `IncomingData` can be immutable (and you don't have any more `while(true)` loops)

t-keazirian replied:
Yes, I can remove all `while(true)` loops for the simple-server (only sending one message).

tpetersen0308 commented on TKeazirian.HTTPServer/Program.cs:
```diff
@@ -1,11 +1,63 @@
-        var hour = DateTime.Now.Hour;
-        var greeting = Greeting.GetGreeting(hour);
-        Console.WriteLine(greeting);
+        byte[] bytes = new byte[1024];
```
What's this for? Why `1024`? Instantiating this so far from where it gets used makes it difficult for me to understand its purpose as the reader

t-keazirian replied:
Good point.  From research, my understanding is it needs to indicate a certain number of bytes to receive?

MarieRobinson replied:
yes.  in C# an array is defined with a fixed size.  so this array will allow for up to 1024 bytes.

tpetersen0308 commented on TKeazirian.HTTPServer/Program.cs:
```diff
@@ -1,11 +1,63 @@
+
+        IPHostEntry ipHostInfo = Dns.GetHostEntry(Dns.GetHostName());
+        IPAddress ipAddress = ipHostInfo.AddressList[0];
+        IPEndPoint localEndPoint = new IPEndPoint(ipAddress, 11000);
```
Is `11000` the port? Why `11000`? Could the port be something the user gets to decide?

t-keazirian replied:
Hmm...I assume it is possible but I'm not sure how (yet) it could be done so the user can decide...

I chose port 11000 sort of randomly, from examples I've seen.  I also tried it `1234`.

tpetersen0308 commented on TKeazirian.HTTPServer/Program.cs:
```diff
@@ -23,18 +23,31 @@ private static void StartListening()
-            IncomingData += Encoding.ASCII.GetString(bytes, 0, bytesReceived);
 
-            Console.WriteLine($"Text received: {IncomingData}");
+            request += Encoding.ASCII.GetString(bytes, 0, bytesReceived);
```
Looks like this is never actually reassigned. Can this just be something like
`string request = Encoding.ASCII...`

t-keazirian replied:
good catch - fixed!

t-keazirian commented:
> Seems like you've got the basic echo behavior working. I left some comments, but I'm aware that you've mostly been focused on getting the implementation from the microsoft docs working. I think there's some refactoring to be done for readability/code hygiene. The `StartListening` method is pretty big--I'm wondering how that might be broken up in to smaller (perhaps test driven?) units that clarify the control flow.

I added in a test and refactored some methods into their own class.  Let me know if that is what you had in mind!
